nvbios/nvbios.c

```diff
@@ -1802,39 +1804,71 @@ int main(int argc, char **argv) {
 				printf("\n");
 			}
 		} else {
+			int min_uv, max_uv;
 			if (is_pwm_based) {
 				printf("-- Mode PWM, acceptable range [%d, %d] µV, frequency %d kHz, base voltage %d µV (unk = %d), range %d µV --\n\n",
 					le32(start+10), le32(start+14), le32(start+5) / 1000000,
 					le32(start+18) & 0xfffff,
 					le32(start+18) >> 20, le32(start+22));
-			} else {
+			} else if (is_header_based) {
 				step_uv = le16(start+22);
+				min_uv = le32(start+10);
+				max_uv = le32(start+14);
 				printf("-- Mode GPIO, Base voltage %d µV, voltage step %d µV, acceptable range [%d, %d] µV --\n\n",
-					volt_uv, step_uv, le32(start+10), le32(start+14));
+					volt_uv, step_uv, min_uv, max_uv);
+			} else {
+				printf("-- Mode GPIO --\n\n");
 			}
 
 			start += header_length;
 
+			/* header based ones get their voltage generate out of the header */
+			if (is_header_based && !is_pwm_based) {
+				int cur_volt = volt_uv;
+				for (i = 0;; ++i, cur_volt += step_uv) {
+					if (step_uv > 0 && cur_volt > max_uv)
+						break;
+
+					if (step_uv > 0 && cur_volt < min_uv)
+						continue;
+
+					if (step_uv < 0 && cur_volt < min_uv)
+						break;
+
+					if (step_uv < 0 && cur_volt > max_uv)
+						continue;
+
+					printf("-- Vid %d, voltage %d µV --\n", i, cur_volt);
+					for (int j = 0; j < nr_vidtag; j++) {
+						if (!(mask & (1 << j)))
+							continue;
+						printf("-- GPIO tag 0x%x(VID) data (logic %d) --\n", vidtag[j], (!!(i & (1 << j))));
+					}
+					printf("\n");
+				}
+			}
+
 			for (i = 0; i < entry_count; i++) {
 				volt = le32(start) & 0x001fffff;
 
-				if (is_pwm_based) {
-					printf("-- Vid %d, voltage %d µV (unk = %d) --\n", i, volt, le32(start) >> 21);
+				if (is_pwm_based || is_header_based) {
+					int v = le32(start);
+					if (v)
+						printf("-- entry %d, unk = %x --\n", i, v);
 				} else {
-					printf("-- Vid %d, voltage %d µV/%d µV (unk = %d) --\n", i, volt_uv, volt, le32(start) >> 21);
-					volt_uv += step_uv;
+					int vid = le32(start) >> 23 & 0xff;
+					/* start >> 21 & 0x3 can be still part of the voltage*/
+					printf("-- Vid %d, voltage %d µV (unk21 = %d, unk31 = %d) --\n",
+						i, volt, le32(start) >> 21 & 0x3, le32(start) >> 31 & 0x1);
 					/* List the gpio tags assosiated with each voltage id */
 					int j;
 					for (j = 0; j < nr_vidtag; j++) {
-						if (!(mask & (1 << j))) {
-						/*	printf("-- Voltage unused/overridden by voltage mask --\n");*/
+						if (!(mask & (1 << j)))
 							continue;
-						}
-						printf("-- GPIO tag 0x%x(VID) data (logic %d) --\n", vidtag[j], (!!(i & (1 << j))));
+						printf("-- GPIO tag 0x%x(VID) data (logic %d) --\n", vidtag[j], (!!(vid & (1 << j))));
```

Comment:
This looks like an unrelated fix, am I right?

Please at least document it in the commit message or by adding a comment.